sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java

```diff
@@ -105,72 +115,84 @@ Object decodeUTF8Text(byte[] bytes) {
-                         * ^
-                         */
-                        valIdx = i;
-                    }
-                    break;
-                case '}':
-                    level--;
-                    if (level == 0) {
-                        /* {"10",NULL,"20"}
-                         *           |    |
-                         *           |    i
-                         *           valIdx
-                         */
-                        byte firstValueByte = bytes[valIdx + 1];
-                        if (firstValueByte == '"') {
-                            byte[] innerBytes = Arrays.copyOfRange(bytes, valIdx + 2 , i - 1);
-                            String s = new String(innerBytes, StandardCharsets.UTF_8);
-                            s = s.replace("\\\"", "\"");
-                            s = s.replace("\\\\", "\\");
-                            values.add(innerType.decodeUTF8Text(s.getBytes(StandardCharsets.UTF_8)));
-                        } else if (firstValueByte == 'N') {
-                            values.add(null);
+                    if (i == 0 || bytes[i - 1] != '"') {
+                        if (i > 0) {
+                            // n-dimensions array -> call recursively
+                            List<Object> nestedObjects = new ArrayList<>();
+                            i = decodeUTF8Text(bytes, i + 1, endIdx, nestedObjects);
+                            valIdx = i;
+                            objects.add(nestedObjects.toArray());
+                        } else {
+                            // 1-dimension array -> call recursively
+                            i = decodeUTF8Text(bytes, i + 1, endIdx, objects);
                         }
+                    } else {
+                        // Start of JSON object
+                        valIdx = i - 1;
+                        jsonObject = true;
                     }
                     break;
                 case ',':
-                    if (level == 1) {
-                        /* {"10",NULL,"20"}
-                         * ||   |
-                         * ||   i
-                         * |firstValueByte
-                         * valIdx
-                         */
-                        byte firstValueByte = bytes[valIdx + 1];
-                        if (firstValueByte == '"') {
-                            byte[] innerBytes = Arrays.copyOfRange(bytes, valIdx + 2 , i - 1);
-                            String s = new String(innerBytes, StandardCharsets.UTF_8);
-                            s = s.replace("\\\"", "\"");
-                            s = s.replace("\\\\", "\\");
-                            values.add(innerType.decodeUTF8Text(s.getBytes(StandardCharsets.UTF_8)));
-                        } else if (firstValueByte == 'N') {
-                            values.add(null);
-                        }
-                        valIdx = i;
+                    if (!jsonObject) {
```

Comment:
I have it there since normally you parse something like: {val1, val2, val3}
So the ',' is in between the curly brackets.